### ISSUE TITLE
Introduce a Unit type

### ISSUE THREAD

**Issue by Abnaxos** [vavr-io/vavr#2396]
Introduce a Unit type like Scala has. This could look something like this:

```java
public final class Unit implements Serializable {

  private static final long serialVersionUID = 2019_03_20_00L;

  private static final Unit INSTANCE = new Unit();

  private Unit() {
  }

  public static Unit unit() {
    return INSTANCE;
  }

  @Override
  public String toString() {
    return "<Unit>";
  }

  private Object readResolve() throws ObjectStreamException {
    return unit();
  }

}
```

Java has `Void`, but because by definition, `Void` cannot have any instances which requires us to return null e.g. from a `Function<Void>`. I avoid null like the plague and setup my IDE so that it warns me about a returning null from a function. `return unit()` would be much  better.

Unit on the other hand, is by definition the type of a value that has only one value. Therefore, it's a better substitute for a function with no return value and avoids returning null. It also allows to use the return value of void functions in other cases, e.g. `Try<Unit>`, `Future<Unit>`.

And it looks scalaish which is a good thing.

Optionaly also add utility conversion functions:

- `Runnable` → `Function0<Unit>`
- `Callable` → `CheckedFunction0<Unit>`
 
This would supersede https://github.com/vavr-io/vavr/issues/71.

**Comment by Abnaxos** [comment]
Also related to #2397: `None<Void>` would be a viable alternative to `Unit`, but certainly not `Some<Void>`.

**Comment by danieldietrich** [comment]
This question has been also already discussed. I recognize that it is asked only by developers who use JSR-305 Nonnull. It would be only a workaround for the IDE tooling.

Scala has Unit, but it is completely different - it is part of the native type system and deeply integrated into the language. In Java/VAVR, Unit would be an ugly workaround that stands on its own.

If I would introduce Unit, I would be with us forever. Instead I will wait for primitive generics. Then we have native Java support: `Function<void, void>` = `Runnable`.

**Comment by danieldietrich** [comment]
Related: #1919, #2261, #2117, #1701

**Comment by danieldietrich** [comment]
Please don't feel bad because I close so many issues. We can still discuss them here. But I need to manage this open source project somehow. There are many questions (see closed issues).

I love TypeScript. It has one of the most sophisticated (and in fact Turing complete) type systems that exist today. Especially union and intersection types are interesting. That would also be a solution for Java to make the use of null more transparent (by making null a type).

```
// = 'null'
typeof null;

// = 'undefined'
typeof null;

// f(null) only works if we make it explicit on the type system
function f<T>(param: T | null) { ... }
```

**Comment by Abnaxos** [comment]
Yes. I chose `Unit` because it looks nice, scalaish and it's short.See #2397, this was the important one.

`None<Void>` would be a good javaish alternative or just `Void`. Actually, with the current API, this can be left to the user. Closing is just fine.

Currently, with a 1.0 branch active, I tend report issues or thoughts rather *now* than later. After 1.0 it will be much harder to introduce potentially breaking changes.
